### PR TITLE
prevent panics in galley

### LINKED ISSUE #22478: Galley panic on analysis
https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/22467/integ-galley-k8s-tests_istio/10301

### LINKED ISSUE #30133: status: panic in CRD discovery
```
2021-01-15T22:38:58.559225Z     info    source  Beginning CRD Discovery, to figure out resources that are available...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x251ee41]

goroutine 3413639 [running]:
istio.io/istio/galley/pkg/config/source/kube/apiserver/status.(*state).applyMessages(0x0, 0xc01d173b08)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/status/state.go:123 +0x41
istio.io/istio/galley/pkg/config/source/kube/apiserver/status.(*ControllerImpl).Report(0xc02e2636b0, 0xc022000000, 0x11, 0x20)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/status/controller.go:144 +0x388
istio.io/istio/galley/pkg/config/source/kube/apiserver.(*Source).Update(0xc02d4c9b80, 0xc022000000, 0x11, 0x20)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/source.go:251 +0x5d
istio.io/istio/galley/pkg/config/processing/snapshotter.(*AnalyzingDistributor).analyzeAndDistribute(0xc007c14600, 0xc029c56e40, 0xc02db52910, 0x7, 0xc01368e2a0, 0xc0212ec5d0)
        istio.io/istio/galley/pkg/config/processing/snapshotter/analyzingdistributor.go:166 +0x2ca
created by istio.io/istio/galley/pkg/config/processing/snapshotter.(*AnalyzingDistributor).Distribute
        istio.io/istio/galley/pkg/config/processing/snapshotter/analyzingdistributor.go:139 +0x23c
```

Not sure what triggered it, just left my cluster alone for a while

### LINKED ISSUE #30160: status: "panic: session.stop: Invalid state: terminating"
```
=== RUN   TestAllNamespaces
2021-01-18T14:49:08.073646Z	info	tf	=== BEGIN: Test: 'pilot[TestAllNamespaces]' ===
panic: session.stop: Invalid state: terminating
goroutine 5890 [running]:
istio.io/istio/galley/pkg/config/processing.(*session).stop(0xc0015c4a80)
	/home/prow/go/src/istio.io/istio/galley/pkg/config/processing/session.go:166 +0x45c
istio.io/istio/galley/pkg/config/processing.(*Runtime).run(0xc00043c070, 0x0, 0xc0006ac360)
	/home/prow/go/src/istio.io/istio/galley/pkg/config/processing/runtime.go:148 +0x40b
created by istio.io/istio/galley/pkg/config/processing.(*Runtime).Start
	/home/prow/go/src/istio.io/istio/galley/pkg/config/processing/runtime.go:88 +0x105
```

Encountered in CI. This did NOT have https://github.com/istio/istio/pull/29275 - although the galley code is not touched by that PR

### PR DESCRIPTION
**Pull request by therealmitchconnors**
Fixes #22478
Fixes #30160 
Fixes #30133

### CODE REVIEW

**Comment by therealmitchconnors** [comment]
Accidentally closed due to awkwardness on mobile interface.  Reopening.

**Comment by howardjohn** [comment]
Still getting a crash:
```
goroutine 11926097 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2a8cb20, 0x4f767a0)
        k8s.io/apimachinery@v0.20.1/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        k8s.io/apimachinery@v0.20.1/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2a8cb20, 0x4f767a0)
        runtime/panic.go:969 +0x1b9
istio.io/istio/galley/pkg/config/source/kube/apiserver/status.(*state).setObserved(0x0, 0xc0024b6bd0, 0x30, 0xc01fe7d030, 0xf, 0xc01fe7d010, 0xa, 0xc01fe7d008, 0x7, 0x0, ...)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/status/state.go:88 +0x3e
istio.io/istio/galley/pkg/config/source/kube/apiserver/status.(*ControllerImpl).UpdateResourceStatus(0xc003e4c9c0, 0xc0024b6bd0, 0x30, 0xc01fe7d030, 0xf, 0xc01fe7d010, 0xa, 0xc01fe7d008, 0x7, 0x2a580a0, ...)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/status/controller.go:111 +0xf8
istio.io/istio/galley/pkg/config/source/kube/apiserver.(*watcher).handleEvent(0xc02dda4960, 0x2, 0x2fc45c0, 0xc01a211180)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/watcher.go:130 +0x407
istio.io/istio/galley/pkg/config/source/kube/apiserver.(*watcher).start.func2(0x2fc45c0, 0xc01f430bf8, 0x2fc45c0, 0xc01a211180)
        istio.io/istio/galley/pkg/config/source/kube/apiserver/watcher.go:80 +0x9a
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
        k8s.io/client-go@v0.20.1/tools/cache/controller.go:238
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
        k8s.io/client-go@v0.20.1/tools/cache/shared_informer.go:775 +0x1c5
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc01636f760)
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0040a3f60, 0x337d160, 0xc0335f72c0, 0x29f5c01, 0xc027498cc0)
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc01636f760, 0x3b9aca00, 0x0, 0x1, 0xc027498cc0)
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(...)
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:90
k8s.io/client-go/tools/cache.(*processorListener).run(0xc020a50f80)
        k8s.io/client-go@v0.20.1/tools/cache/shared_informer.go:771 +0x95
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc01a1d8920, 0xc00c4b9420)
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:73 +0x51
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
        k8s.io/apimachinery@v0.20.1/pkg/util/wait/wait.go:71 +0x65
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2598e5e]
```

File: galley/pkg/config/analysis/local/source.go [hunk 1]
```diff
@@ -78,10 +79,13 @@ func (ph *precedenceHandler) Handle(e event.Event) {
 // For each collection, we want to only send this once, after all upstream sources have sent theirs.
 func (ph *precedenceHandler) handleFullSync(e event.Event) {
 	col := e.Source.Name()
+	ph.src.countMu.Lock()
 	ph.src.expectedCounts[col]--
 	if ph.src.expectedCounts[col] > 0 {
+		ph.src.countMu.Unlock()
```
Comment (esnible) [inline galley/pkg/config/analysis/local/source.go:85]: Nit: Code with two Unlocks takes brainpower to review, and have a strong preference for

```
ph.src.countMu.Lock()
ph.src.expectedCounts[col]--
shallHandle := ph.src.expectedCounts[col] > 0
ph.src.countMu.Unlock()
if (shallHandle) { ph.src.handler.Handle(e) }
```

(If we do it your way in the rest of Istio feel free to ignore this comment.)

File: galley/pkg/config/source/kube/apiserver/source.go [hunk 2]
```diff
@@ -193,6 +193,10 @@ func (s *Source) startWatchers() {
 	scope.Source.Info("Creating watchers for Kubernetes CRDs")
 	s.watchers = make(map[collection.Name]*watcher)
 	for i, r := range resources {
+		if s.provider == nil {
```
Comment (esnible) [inline galley/pkg/config/source/kube/apiserver/source.go:196]: Need a Mutex to guard against `s.provider` going nil between this line and the use of it to create `a` on line 200?

File: galley/pkg/config/source/kube/apiserver/status/controller.go [hunk 3]
```diff
@@ -141,6 +141,10 @@ func (c *ControllerImpl) Report(messages diag.Messages) {
 		msgs.Add(origin, m)
 	}
 
+	if c.state == nil {
```
Comment (esnible) [inline galley/pkg/config/source/kube/apiserver/status/controller.go:144]: Same comment as before.  Putting the if right before the use makes this more reliable but don't we need to guard against it going nil between this line and `applyMessages()`?

**Review by esnible — CHANGES_REQUESTED** [review]

File: galley/pkg/config/source/kube/apiserver/status/controller.go (hunk 3 above)
Comment (therealmitchconnors) [reply galley/pkg/config/source/kube/apiserver/status/controller.go:144]: This function is already only called while the `mu` lock is held, so it cannot be set to nil between these lines.

File: galley/pkg/config/source/kube/apiserver/source.go (hunk 2 above)
Comment (therealmitchconnors) [reply galley/pkg/config/source/kube/apiserver/source.go:196]: This function is already only called while the `mu` lock is held, so it cannot be set to nil between these lines.

File: galley/pkg/config/source/kube/apiserver/status/controller.go (hunk 3 above)
Comment (therealmitchconnors) [reply galley/pkg/config/source/kube/apiserver/status/controller.go:148]: will add rlock here

**Review by therealmitchconnors — COMMENTED** [review]
I have added read locks to prevent known panic conditions, and added guard rails in pilot to prevent any future panics from impacting our ability to serve xds.

**Review by esnible — APPROVED** [review]
The code looks good with the changes you made.  I didn't do an audit but don't see any flaws I recognize.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**